framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,854 @@
+      elif all(opt[var] == old[var] for var in self.toBeSampled):
+        # this is the classic "same point" trap; we accept the same point, and check convergence later
+        acceptable = 'accepted'
+      else:
+        acceptable = self._checkForImprovement(optVal, oldVal)
+    else: # no history
+      # if first sample, simply assume it's better!
+      acceptable = 'first'
+      old = None
+    self._acceptHistory[traj].append(acceptable)
+    self.raiseADebug(' ... {a}!'.format(a=acceptable))
+    return acceptable, old
+
+  def _checkForImprovement(self, new, old):
+    """
+      Determine if the new value is sufficient improved over the old.
+      @ In, new, float, new optimization value
+      @ In, old, float, previous optimization value
+      @ Out, improved, bool, True if "sufficiently" improved or False if not.
+    """
+    # TODO could this be a base RavenSampled class?
+    improved = self._acceptInstance.checkImprovement(new, old)
+    return 'accepted' if improved else 'rejected'
+
+  def _updateConvergence(self, traj, new, old, acceptable):
+    """
+      Updates convergence information for trajectory
+      @ In, traj, int, identifier
+      @ In, acceptable, str, condition of point
+      @ Out, converged, bool, True if converged on ANY criteria
+    """
+    ## NOTE we have multiple "if acceptable" trees here, as we need to update soln export regardless
+    if acceptable == 'accepted':
+      self.raiseADebug('Convergence Check for Trajectory {}:'.format(traj))
+      # check convergence
+      converged, convDict = self.checkConvergence(traj, new, old)
+    else:
+      converged = False
+      convDict = dict((var, False) for var in self._convergenceInfo[traj])
+    self._convergenceInfo[traj].update(convDict)
+    return converged
+
+  def _updatePersistence(self, traj, converged, optVal):
+    """
+      Update persistence tracking state variables
+      @ In, traj, int, identifier
+      @ In, converged, bool, convergence check result
+      @ In, optVal, float, new optimal value
+      @ Out, None
+    """
+    # update persistence
+    if converged:
+      self._convergenceInfo[traj]['persistence'] += 1
+      self.raiseADebug('Trajectory {} has converged successfully {} / {} time(s)!'.format(traj, self._convergenceInfo[traj]['persistence'], self._requiredPersistence))
+      if self._convergenceInfo[traj]['persistence'] >= self._requiredPersistence:
+        self._closeTrajectory(traj, 'converge', 'converged', optVal)
+    else:
+      self._convergenceInfo[traj]['persistence'] = 0
+      self.raiseADebug('Resetting convergence for trajectory {}.'.format(traj))
+
+  def _addToSolutionExport(self, traj, rlz, acceptable):
+    """
+      Contributes additional entries to the solution export.
+      @ In, traj, int, trajectory which should be written
+      @ In, rlz, dict, collected point
+      @ In, acceptable, bool, acceptability of opt point
+      @ Out, toAdd, dict, additional entries
```

Comment:
The keys are added dynamically; they are not known a priori, therefore the contents cannot be explicitly listed. The method is very short, so observing the way the dictionary is dynamically populated should be straightforward, right?

Reply:
The thing that I want to know is the type of the values of the dictionary. float, list, array? If this information is provided, I can easily reuse/overwrite this method without checking the source code.

Reply:
I don't know what those types will be ahead of run time, I pass this method to other entities to fill values. Possible inclusions are anything that can go in a realization, as explained by `Contributes additional entries to the solution export`. This can potentially be strings, floats, integers, or histories of any of the above, or even N-dimensional data. I don't think it is the task of this method to determine what acceptable realization entries are; that is a task for the DataObject.